job_creator.php

```diff
@@ -737,4 +640,76 @@ public function createJson(string $yml): string
         $json = str_replace("\n", '', $json);
         return trim($json);
     }
-}
\ No newline at end of file
+
+    public function parseRepositoryMetadata()
+    {
+        $this->repoData = MetaData::getMetaDataForRepository($this->githubRepository, true);
+        $this->lockSteppedRepos = MetaData::getMetaDataForLocksteppedRepos();
+    }
+
+    public function getCleanedBranch(string $branch): string
+    {
+        // e.g. pulls/4.10/some-bugfix or pulls/4/some-feature
+        // for push events to the creative-commoners account
+        if (preg_match('#^pulls/([0-9]+(\.[0-9]+)*)/#', $branch, $matches)) {
+            return $matches[1];
+        }
+        // fallback to parent branch if available
+        if (
+            !$this->branchIsSemver($branch) &&
+            $this->parentBranch &&
+            $this->branchIsSemver($this->parentBranch)
+        ) {
+            return $this->parentBranch;
+        }
+        return $branch;
+    }
+
+    private function branchIsSemver(string $branch): bool
+    {
+        return preg_match('/^[0-9]+(\.[0-9]+)*$/', $branch);
+    }
+
+    private function needsInstallerVersion(): bool
+    {
+        if (in_array($this->repoName, FORCE_INSTALLER_REPOS)) {
+            return true;
+        }
+        if (in_array($this->repoName, NO_INSTALLER_REPOS)) {
+            return false;
+        }
+        // We shouldn't try to infer the installer version for regular repositories
+        // that we don't know anything about
+        if (empty($this->repoData)) {
+            if (!file_exists($this->composerJsonPath)) {
+                return false;
+            }
+            $json = $this->getComposerJsonContent();
+            $silverstripeRepoTypes = [
+                'silverstripe-vendormodule',
+                'silverstripe-module',
+                'silverstripe-recipe',
+                'silverstripe-theme',
+            ];
+            return isset($json->type) && in_array($json->type, $silverstripeRepoTypes);
+        }
+        if ($this->repoData['type'] === 'recipe') {
+            return false;
+        }
+        return true;
+    }
+
+    private function getComposerJsonContent(): ?stdClass
+    {
+        if (!$this->composerJsonContent) {
+            if (!file_exists($this->composerJsonPath)) {
+                return null;
+            }
+            $this->composerJsonContent = json_decode(file_get_contents($this->composerJsonPath));
+            if ($this->composerJsonContent === null) {
+                throw new RuntimeException('Could not decode composer.json - last error was: ' . json_last_error_msg());
+            }
+        }
+        return $this->composerJsonContent;
+    }
```

Comment:
Previously we were decoding the file multiple times at different stages - now we just decode it once, whenever the first time is that we need it, and use that same decoded result every time.